server/src/main/java/io/crate/execution/jobs/SharedShardContexts.java

```diff
@@ -102,9 +101,8 @@ public SharedShardContext getOrCreateContext(ShardId shardId) throws IndexNotFou
         SharedShardContext sharedShardContext = allocatedShards.get(shardId);
         if (sharedShardContext == null) {
             IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
-            sharedShardContext = new SharedShardContext(indexService, shardId, readerId, wrapSearcher);
+            sharedShardContext = new SharedShardContext(indexService, shardId, shardId.hashCode(), wrapSearcher);
```

Comment:
If these are always tied to a shard, and there is only a single context per shard for a given job, then do we even need the separate reader ID?

Reply:
Thank you as you said it looks to me that we can eliminate reader ID completely. @mfussenegger could you clarify if this makes sense?

Reply:
In the fetch case the readerId is used to encode into a `FetchId` which shard is used, so that we can lookup a document from the correct shard. See:

- https://github.com/crate/crate/blob/b2f20a38f78b7f3d6b40b6077ad6f2b933429b37/server/src/main/java/io/crate/planner/ReaderAllocations.java#L41-L70
- https://github.com/crate/crate/blob/b2f20a38f78b7f3d6b40b6077ad6f2b933429b37/server/src/main/java/io/crate/execution/engine/fetch/FetchTask.java#L293-L297
- https://github.com/crate/crate/blob/b2f20a38f78b7f3d6b40b6077ad6f2b933429b37/server/src/main/java/io/crate/execution/engine/fetch/FetchId.java#L39


What'd be more interesting is how the increments can lead to duplicates. The whole code section is under the assumption that the preparation phase is run single threaded. See also https://github.com/crate/crate/pull/5248

If that assumption is no longer true - maybe due to https://github.com/crate/crate/pull/10373, or if the SharedShardContexts is accessed elsewhere, we probably need to change more than just the readerId.

Reply:
Thank you for the pointers, I think it is due to https://github.com/crate/crate/pull/10373 which changes `ShardCollectSource.getIterator()` to return `future` type.

https://github.com/crate/crate/blob/a177d69f77d70ee3757533b88b5bb7c8b160d175/server/src/main/java/io/crate/execution/engine/collect/sources/ShardCollectSource.java#L435-L441
When more than one `awaitShardSearchActive()` completes, it would cause a race for a reader ID.

Reply:
I tried to sequentialize the calls to `SharedShardContexts.getOrCreateContext()` which is the only place that assigns reader ids to `SharedShardContext`, 3da107309ef10242258d0c9f6647a4490b0bdc02.

Reply:
Hi @mfussenegger could you have a look if the fix is ok before I start adding tests? Now `SharedShardContexts.getOrCreateContext()` is only called from two places.

Reply:
I think I'd prefer adding synchronization to the methods again. But first we should ensure that we've identified the real problem. Did you do that?

Reply:
I think the real problem is https://github.com/crate/crate/pull/15520#discussion_r1480543804.

For example, If I attach a logger:
```
    public SharedShardContext getOrCreateContext(ShardId shardId) throws IndexNotFoundException {
        LOGGER.info("Begin getOrCreateContext : " + Thread.currentThread());
        SharedShardContext sharedShardContext = allocatedShards.get(shardId);
        if (sharedShardContext == null) {
            IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
            sharedShardContext = new SharedShardContext(indexService, shardId, readerId, wrapSearcher);
            allocatedShards.put(shardId, sharedShardContext);
            readerId++;
        }
        LOGGER.info("End getOrCreateContext : " + Thread.currentThread());
        return sharedShardContext;
    }
```
I can observe two thread being interleaved:
```
[2024-02-07T09:20:43,629][INFO ][i.c.e.j.SharedShardContexts] [Grand Parpaillon] Begin getOrCreateContext : Thread[#356,cratedb[Grand Parpaillon][listener][T#3],5,main]
[2024-02-07T09:20:43,629][INFO ][i.c.e.j.SharedShardContexts] [Grand Parpaillon] Begin getOrCreateContext : Thread[#354,cratedb[Grand Parpaillon][listener][T#1],5,main]
[2024-02-07T09:20:43,629][INFO ][i.c.e.j.SharedShardContexts] [Grand Parpaillon] End getOrCreateContext : Thread[#356,cratedb[Grand Parpaillon][listener][T#3],5,main]
[2024-02-07T09:20:43,629][INFO ][i.c.e.j.SharedShardContexts] [Grand Parpaillon] End getOrCreateContext : Thread[#354,cratedb[Grand Parpaillon][listener][T#1],5,main]
```
right before the exception is thrown.

Reply:
Thank you, reverted.